app/src/main/kotlin/com/wire/android/ui/userprofile/UserProfileViewModel.kt

```diff
@@ -31,28 +29,69 @@ import kotlinx.coroutines.withContext
 
     init {
-        // TODO: here we should have a loading state as the first initial state of the screen
         viewModelScope.launch {
-            getSelf().collect {
-                userProfileState = SelfUserProfileState(
-                    status = UserStatus.AVAILABLE,
-                    fullName = it.name.orEmpty(),
-                    userName = it.handle.orEmpty(),
-                    teamName = it.team,
-                    // Add some mocked team
-                    otherAccounts = listOf(
-                        OtherAccount("someId", "", it.name.orEmpty(), "Wire Swiss GmbH"),
-                        OtherAccount("someId", "", "B. A. Baracus", "The A-Team"),
+            fetchSelfUser()
+        }
+    }
+
+    private suspend fun fetchSelfUser() {
+        viewModelScope.launch {
+            getSelf().collect { selfUser ->
+                with(selfUser) {
+                    // Load user avatar raw image data
+                    completePicture?.let { updateUserAvatar(it) }
+
+                    // Update user data state
+                    userProfileState = SelfUserProfileState(
+                        status = UserStatus.AVAILABLE,
+                        fullName = name.orEmpty(),
+                        userName = handle.orEmpty(),
+                        teamName = team,
+
+                        // TODO: remove mocked team when other accounts functionality is ready
+                        otherAccounts = listOf(
+                            OtherAccount("someId", "", name.orEmpty(), "Wire Swiss GmbH"),
+                            OtherAccount("someId", "", "B. A. Baracus", "The A-Team"),
+                        )
                     )
+                }
+            }
+        }
+    }
+
+    private fun showErrorMessage() {
+        userProfileState = userProfileState.copy(errorMessageCode = GENERIC_DOWNLOAD_USER_INFO_ERROR)
+    }
+
+    private fun showLoadingAvatar(show: Boolean) {
+        userProfileState = userProfileState.copy(isAvatarLoading = show)
+    }
+
+    private fun updateUserAvatar(avatarAssetId: UserAssetId) {
+        // We try to download the user avatar on a separate thread so that we don't block the display of the user's info
+        viewModelScope.launch {
+            try {
+                showLoadingAvatar(true)
+                userProfileState = userProfileState.copy(
+                    avatarAssetByteArray = (getPublicAssetUseCase(avatarAssetId) as PublicAssetResult.Success).asset
                 )
+
+                // Update avatar asset id on user data store
+                // TODO: obtain the asset id through a useCase once we also store assets ids
+                dataStore.updateUserAvatarAssetId(avatarAssetId)
+            } catch (e: ClassCastException) {
+                // Show error snackbar if avatar download fails
+                showErrorMessage()
+            } finally {
+                showLoadingAvatar(false)
             }
         }
```

Comment:
I just now remember that while I was working on it the use cases do not uses the DIspatcher.IO so I would suggest to use it here instead so that we do not freeze the UI